## notebook to combine all the data, from CAMS, ERA5, feature engineering and others

In [1]:
## conda environment Weather_Prediction

import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import geopandas as gpd
import xarray as xr
pd.set_option('display.max_columns', None)

## Before we merge, lets find out counties whose shapefile does not exist, and those counties whose shapefiles does exist but the life expectancy data doesnot exist


In [2]:
## Lets load the county shapefile

SHAPE_PATH = os.path.join('../Shapefiles','county_shapefiles','2010_county_shapefile','gz_2010_us_050_00_500k.shp')
county_gdf = gpd.read_file(SHAPE_PATH)

## lets create the fips column in the dataframe which is the sum of state and county both as dtype string.
## Note: not a numerical sum
## This step is necessary because there can be several counties with idential names. 

county_gdf['fips']=county_gdf['STATE'] + county_gdf['COUNTY']

## convert the data type
county_gdf['fips']=county_gdf['fips'].astype(str).astype(int)

county_gdf = county_gdf.drop([ 'GEO_ID','CENSUSAREA','COUNTY','LSAD'], axis=1)
county_gdf

STATE           NAME                                           geometry  \
0       01       Cleburne  POLYGON ((-85.38872 33.91304, -85.38088 33.873...   
1       01         Coffee  POLYGON ((-86.03044 31.61894, -86.00408 31.619...   
2       01          Coosa  POLYGON ((-86.00928 33.10164, -86.00917 33.090...   
3       01      Covington  POLYGON ((-86.34851 30.99434, -86.35023 30.994...   
4       01       Crenshaw  POLYGON ((-86.14699 31.68045, -86.14711 31.663...   
...    ...            ...                                                ...   
3216    72  San Sebastián  POLYGON ((-66.90748 18.25314, -66.90739 18.253...   
3217    72   Santa Isabel  POLYGON ((-66.37968 17.94398, -66.38029 17.943...   
3218    72       Toa Baja  MULTIPOLYGON (((-66.13957 18.46232, -66.13956 ...   
3219    72  Trujillo Alto  POLYGON ((-66.02917 18.37590, -66.02828 18.376...   
3220    72          Yauco  POLYGON ((-66.85229 17.95500, -66.85280 17.955...   

       fips  
0      1029  
1      1031  
2      1037  
3      1039  
4      1041  
...     ...  
3216  72131  
3217  72133  
3218  72137  
3219  72139  
3220  72153  

[3221 rows x 4 columns]

In [3]:
## gather only state fips of USA not its territories such as guam, puerto rico

county_gdf['STATE']=county_gdf['STATE'].astype(str).astype(int) ## convert state dtype to int
county_gdf=county_gdf.loc[(county_gdf['STATE'] < 60)]
county_gdf

STATE        NAME                                           geometry  \
0         1    Cleburne  POLYGON ((-85.38872 33.91304, -85.38088 33.873...   
1         1      Coffee  POLYGON ((-86.03044 31.61894, -86.00408 31.619...   
2         1       Coosa  POLYGON ((-86.00928 33.10164, -86.00917 33.090...   
3         1   Covington  POLYGON ((-86.34851 30.99434, -86.35023 30.994...   
4         1    Crenshaw  POLYGON ((-86.14699 31.68045, -86.14711 31.663...   
...     ...         ...                                                ...   
3138     56    Niobrara  POLYGON ((-104.05298 42.85955, -104.05286 42.7...   
3139     56      Platte  POLYGON ((-104.77417 42.60996, -104.76422 42.6...   
3140     56  Sweetwater  POLYGON ((-109.05008 41.00066, -109.17368 41.0...   
3141     56    Washakie  POLYGON ((-107.12892 43.99455, -107.12797 43.9...   
3142     56      Weston  POLYGON ((-104.05518 43.76163, -104.05514 43.7...   

       fips  
0      1029  
1      1031  
2      1037  
3      1039  
4      1041  
...     ...  
3138  56027  
3139  56031  
3140  56037  
3141  56043  
3142  56045  

[3143 rows x 4 columns]

In [4]:
column_values = county_gdf['STATE'].unique()
column_values

array([ 1,  2,  4,  5,  6,  8,  9, 10, 11, 12, 13, 15, 16, 17, 18, 19, 20,
       21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37,
       38, 39, 40, 41, 42, 44, 45, 46, 47, 48, 49, 50, 51, 53, 54, 55, 56])

In [5]:
## The life expectency data

PATH = os.path.join('../life_expectancy_data','LE_2010.csv')
LE_DATA = pd.read_csv(PATH)
LE_DATA

location_name   fips  year  MeanLifeExpectency
0        Autauga County (Alabama)   1001  2010           75.728489
1        Baldwin County (Alabama)   1003  2010           77.826608
2        Barbour County (Alabama)   1005  2010           75.841973
3           Bibb County (Alabama)   1007  2010           73.705432
4         Blount County (Alabama)   1009  2010           75.753407
...                           ...    ...   ...                 ...
3122  Sweetwater County (Wyoming)  56037  2010           77.582422
3123       Teton County (Wyoming)  56039  2010           83.771012
3124       Uinta County (Wyoming)  56041  2010           77.836526
3125    Washakie County (Wyoming)  56043  2010           78.674180
3126      Weston County (Wyoming)  56045  2010           79.037741

[3127 rows x 4 columns]

In [6]:
## the counties that are in life expectency data but the shape file does not exist

county_noshape = LE_DATA.loc[~LE_DATA['fips'].isin(county_gdf['fips'])].copy()
county_noshape

location_name   fips  year  \
80                         Chugach Census Area (Alaska)   2063  2010   
81                    Copper River Census Area (Alaska)   2066  2010   
83                        Kusilvak Census Area (Alaska)   2158  2010   
85                           Kobuk Census Area (Alaska)   2140  2010   
87                Aleutian Islands Census Area (Alaska)   2010  2010   
96          Skagway-Yakutat-Angoon Census Area (Alaska)   2231  2010   
97           Skagway-Hoonah-Angoon Census Area (Alaska)   2232  2010   
102   Prince of Wales-Outer Ketchikan Census Area (A...   2201  2010   
104            Wrangell-Petersburg Census Area (Alaska)   2280  2010   
338                               Dade County (Florida)  12025  2010   
1657                Yellowstone National Park (Montana)  30113  2010   
2404                Oglala Lakota County (South Dakota)  46102  2010   
2417                   Washabaugh County (South Dakota)  46131  2010   
2931                       South Boston City (Virginia)  51780  2010   
2935                      Clifton Forge City (Virginia)  51560  2010   

      MeanLifeExpectency  
80             77.733009  
81             77.733009  
83             69.723907  
85             73.339132  
87             82.631153  
96             80.202830  
97             80.202830  
102            78.068610  
104            78.068610  
338            80.576828  
1657           79.589703  
2404           66.689077  
2417           73.724949  
2931           76.010437  
2935           76.032809

## Note that, among the 3127 total counties, 38 are that of Alaska, 5 are that of Hawaii and as shown above there are 6 counties where the life expectancy data does not exist. In total, 3127-5-6= 3078 counties 

In [7]:
## the counties that are in shapefile data but life expectancy data does not exist

le_data_nodata = county_gdf.loc[~county_gdf['fips'].isin(LE_DATA['fips'])].copy()
le_data_nodata

STATE           NAME                                           geometry  \
244       8       Hinsdale  POLYGON ((-107.21157 37.42296, -107.21787 37.4...   
250       8        Mineral  POLYGON ((-106.71077 37.40423, -106.71077 37.3...   
290       8       San Juan  POLYGON ((-107.92580 37.63979, -107.92613 37.6...   
585      16          Clark  POLYGON ((-112.81416 44.37720, -112.81324 44.3...   
1625     30      Petroleum  POLYGON ((-108.50506 46.74914, -108.63095 46.7...   
1640     30       Treasure  POLYGON ((-107.05209 45.87700, -107.05203 45.8...   
1645     30  Golden Valley  POLYGON ((-109.31590 46.75150, -109.26339 46.7...   
1666     31          Grant  POLYGON ((-101.62042 41.74240, -101.63796 41.7...   
1667     31         Hooker  POLYGON ((-101.41220 42.09211, -101.37545 42.0...   
1669     31          Logan  POLYGON ((-100.71171 41.73976, -100.50060 41.7...   
1673     31         Thomas  POLYGON ((-100.36247 41.73964, -100.49075 41.7...   
1675     31         Arthur  POLYGON ((-101.64019 41.74244, -101.63796 41.7...   
1676     31         Banner  POLYGON ((-104.05283 41.69795, -104.00572 41.6...   
1677     31         Blaine  POLYGON ((-99.96281 41.73959, -100.00041 41.73...   
1695     31      Keya Paha  POLYGON ((-99.67002 42.99791, -99.56928 42.997...   
1700     31           Loup  POLYGON ((-99.68662 41.79254, -99.68660 41.798...   
1701     31      McPherson  POLYGON ((-100.81173 41.39378, -100.84897 41.3...   
1717     31          Hayes  POLYGON ((-100.80002 40.34942, -100.81331 40.3...   
1732     31        Wheeler  POLYGON ((-98.29544 42.08887, -98.29542 42.081...   
1751     32      Esmeralda  POLYGON ((-117.16600 36.97122, -117.24492 37.0...   
1820     35        Harding  POLYGON ((-103.37923 35.49496, -103.37994 35.3...   
1995     38          Slope  POLYGON ((-104.04542 46.28019, -104.04547 46.3...   
2002     38       Billings  POLYGON ((-103.48285 46.62976, -103.55040 46.6...   
2536     48         Kenedy  MULTIPOLYGON (((-97.45309 27.05405, -97.45690 ...   
2537     48           Kent  POLYGON ((-100.51921 32.96293, -100.58364 32.9...   
2538     48           King  POLYGON ((-100.39270 33.83594, -100.32486 33.8...   
2544     48         Loving  POLYGON ((-103.57876 32.00027, -103.32650 32.0...   
2545     48       McMullen  POLYGON ((-98.61082 28.64834, -98.43832 28.648...   
2579     48        Terrell  POLYGON ((-102.29733 29.87519, -102.30138 29.8...   
2696     48        Roberts  POLYGON ((-100.54101 35.82400, -100.54128 35.7...   
2736     48         Borden  POLYGON ((-101.68908 32.58967, -101.68926 32.6...   

       fips  
244    8053  
250    8079  
290    8111  
585   16033  
1625  30069  
1640  30103  
1645  30037  
1666  31075  
1667  31091  
1669  31113  
1673  31171  
1675  31005  
1676  31007  
1677  31009  
1695  31103  
1700  31115  
1701  31117  
1717  31085  
1732  31183  
1751  32009  
1820  35021  
1995  38087  
2002  38007  
2536  48261  
2537  48263  
2538  48269  
2544  48301  
2545  48311  
2579  48443  
2696  48393  
2736  48033

## Lets start combining data

In [8]:
## load the data with county and corresponding variables

county_var=pd.read_pickle('CAMS_79_variables.pkl')
county_var

NAME   fips       u10       v10         d2m         t2m  bcaod550  \
0       Cleburne   1029  0.434079 -0.122697  282.638367  289.807770  0.006281   
1         Coffee   1031  0.277523 -0.034499  284.563934  291.594208  0.006543   
2          Coosa   1037  0.262713 -0.041072  283.402100  290.892761  0.006274   
3      Covington   1039  0.219254  0.026002  284.972931  291.764221  0.006611   
4       Crenshaw   1041  0.260638 -0.015809  284.221710  291.456573  0.006468   
...          ...    ...       ...       ...         ...         ...       ...   
3138    Niobrara  56027  1.226940  0.222518  271.670319  282.334412  0.005382   
3139      Platte  56031  1.541398  0.317140  271.038757  281.676422  0.005196   
3140  Sweetwater  56037  1.430520  0.814952  269.132935  278.168732  0.004792   
3141    Washakie  56043  0.710449  0.532009  270.020416  278.829376  0.005376   
3142      Weston  56045  1.077338  0.246453  271.797424  281.669983  0.005429   

      duaod550       lsm            msl  omaod550           pm1         pm2p5  \
0     0.007998  0.985105  101706.890625  0.077957  1.335776e-08  1.673881e-08   
1     0.012055  0.992762  101687.679688  0.073746  1.180315e-08  1.550131e-08   
2     0.010251  0.984186  101696.562500  0.076223  1.158120e-08  1.503791e-08   
3     0.012473  0.938187  101689.382812  0.074370  1.182135e-08  1.562129e-08   
4     0.011628  0.980219  101690.453125  0.074433  1.146198e-08  1.507656e-08   
...        ...       ...            ...       ...           ...           ...   
3138  0.007533  0.998026  101400.070312  0.054008  6.728600e-09  1.009636e-08   
3139  0.006530  0.997104  101402.078125  0.051276  6.485069e-09  9.121436e-09   
3140  0.003537  0.992391  101611.039062  0.043755  5.239014e-09  6.881977e-09   
3141  0.003680  0.997272  101436.078125  0.050795  5.894271e-09  7.628212e-09   
3142  0.007828  0.998318  101415.359375  0.054489  7.089420e-09  1.112587e-08   

              pm10  ssaod550  suaod550             z             sp    aod469  \
0     2.318515e-08  0.006267  0.078404   2581.377930   98584.976562  0.216394   
1     2.120308e-08  0.008770  0.067104    803.111572  100711.117188  0.204075   
2     2.065578e-08  0.007309  0.073413   1495.930664   99881.492188  0.211330   
3     2.134704e-08  0.009069  0.068461    763.528442  100760.820312  0.207274   
4     2.061333e-08  0.008382  0.068510   1009.594543  100463.648438  0.205725   
...            ...       ...       ...           ...            ...       ...   
3138  1.574939e-08  0.004639  0.025946  14033.678711   85428.031250  0.118499   
3139  1.382634e-08  0.004417  0.025776  15923.485352   83507.078125  0.113475   
3140  1.016216e-08  0.005140  0.023394  21153.457031   78338.398438  0.098355   
3141  1.102624e-08  0.004972  0.022588  17578.353516   81764.500000  0.106893   
3142  1.761679e-08  0.004887  0.025279  13694.173828   85765.351562  0.118809   

        aod550    aod670    aod865   aod1240      tcco   tc_c2h6    tchcho  \
0     0.176908  0.135071  0.092637  0.051929  0.000884  0.000003  0.000004   
1     0.168218  0.130380  0.092329  0.054776  0.000880  0.000003  0.000004   
2     0.173469  0.133421  0.092969  0.053567  0.000888  0.000003  0.000004   
3     0.170984  0.132668  0.094133  0.055998  0.000879  0.000003  0.000004   
4     0.169423  0.131085  0.092480  0.054480  0.000882  0.000003  0.000004   
...        ...       ...       ...       ...       ...       ...       ...   
3138  0.097507  0.075646  0.053522  0.033106  0.000727  0.000003  0.000001   
3139  0.093194  0.072066  0.050698  0.030998  0.000702  0.000003  0.000001   
3140  0.080618  0.062090  0.043318  0.025987  0.000641  0.000003  0.000001   
3141  0.087412  0.067119  0.046584  0.027667  0.000686  0.000003  0.000001   
3142  0.097913  0.076147  0.054096  0.033714  0.000733  0.000003  0.000001   

       tc_h2o2         tc_oh       tc_c5h8    tc_ch4   tc_hno3     tcno2  \
0     0.000008  4.903662e-09  2.851306e-06  0.009726  0.000005  0.000004   
1   

In [9]:
## The life expectency data

PATH = os.path.join('../life_expectancy_data','LE_2010.csv')
LE_DATA = pd.read_csv(PATH)
LE_DATA

location_name   fips  year  MeanLifeExpectency
0        Autauga County (Alabama)   1001  2010           75.728489
1        Baldwin County (Alabama)   1003  2010           77.826608
2        Barbour County (Alabama)   1005  2010           75.841973
3           Bibb County (Alabama)   1007  2010           73.705432
4         Blount County (Alabama)   1009  2010           75.753407
...                           ...    ...   ...                 ...
3122  Sweetwater County (Wyoming)  56037  2010           77.582422
3123       Teton County (Wyoming)  56039  2010           83.771012
3124       Uinta County (Wyoming)  56041  2010           77.836526
3125    Washakie County (Wyoming)  56043  2010           78.674180
3126      Weston County (Wyoming)  56045  2010           79.037741

[3127 rows x 4 columns]

In [10]:
## merge life expectency data with the variables by fips

final_df=LE_DATA.merge(county_var,how='left', on='fips')
final_df=final_df.dropna()
final_df

location_name   fips  year  MeanLifeExpectency  \
0        Autauga County (Alabama)   1001  2010           75.728489   
1        Baldwin County (Alabama)   1003  2010           77.826608   
2        Barbour County (Alabama)   1005  2010           75.841973   
3           Bibb County (Alabama)   1007  2010           73.705432   
4         Blount County (Alabama)   1009  2010           75.753407   
...                           ...    ...   ...                 ...   
3122  Sweetwater County (Wyoming)  56037  2010           77.582422   
3123       Teton County (Wyoming)  56039  2010           83.771012   
3124       Uinta County (Wyoming)  56041  2010           77.836526   
3125    Washakie County (Wyoming)  56043  2010           78.674180   
3126      Weston County (Wyoming)  56045  2010           79.037741   

            NAME       u10       v10         d2m         t2m  bcaod550  \
0        Autauga  0.244891 -0.022881  283.480743  291.070984  0.006298   
1        Baldwin  0.065638  0.113299  286.096985  292.149780  0.006543   
2        Barbour  0.336253 -0.075642  284.153351  291.171997  0.006456   
3           Bibb  0.221962 -0.004751  283.564850  290.792206  0.006248   
4         Blount  0.284933  0.005598  282.847870  289.685669  0.006085   
...          ...       ...       ...         ...         ...       ...   
3122  Sweetwater  1.430520  0.814952  269.132935  278.168732  0.004792   
3123       Teton  1.021023  0.729751  268.077728  274.648254  0.004971   
3124       Uinta  1.101981  0.647358  270.266602  277.642670  0.005171   
3125    Washakie  0.710449  0.532009  270.020416  278.829376  0.005376   
3126      Weston  1.077338  0.246453  271.797424  281.669983  0.005429   

      duaod550       lsm            msl  omaod550           pm1         pm2p5  \
0     0.010663  0.986290  101694.718750  0.076191  1.131367e-08  1.479475e-08   
1     0.013571  0.793853  101686.296875  0.074374  1.082756e-08  1.474964e-08   
2     0.010674  0.987881  101689.679688  0.074089  1.173599e-08  1.520591e-08   
3     0.010553  0.986778  101695.648438  0.077052  1.129947e-08  1.473398e-08   
4     0.008301  0.977253  101711.023438  0.077401  1.163996e-08  1.477842e-08   
...        ...       ...            ...       ...           ...           ...   
3122  0.003537  0.992391  101611.039062  0.043755  5.239014e-09  6.881977e-09   
3123  0.001720  0.982567  101594.773438  0.046533  5.223066e-09  6.534898e-09   
3124  0.002573  0.985319  101640.101562  0.046785  6.811558e-09  8.490385e-09   
3125  0.003680  0.997272  101436.078125  0.050795  5.894271e-09  7.628212e-09   
3126  0.007828  0.998318  101415.359375  0.054489  7.089420e-09  1.112587e-08   

              pm10  ssaod550  suaod550             z             sp    aod469  \
0     2.029277e-08  0.007522  0.072511   1337.441284  100070.937500  0.210819   
1     2.021521e-08  0.010043  0.070549    426.965088  101166.343750  0.211824   
2     2.082824e-08  0.007945  0.067335   1178.228027  100259.171875  0.202453   
3     2.021792e-08  0.007260  0.072814   1267.733643  100154.929688  0.211800   
4     2.039585e-08  0.006082  0.073962   2292.157227   98932.171875  0.209957   
...            ...       ...       ...           ...            ...       ...   
3122  1.016216e-08  0.005140  0.023394  21153.457031   78338.398438  0.098355   
3123  9.327799e-09  0.005876  0.021670  23587.818359   75929.421875  0.098563   
3124  1.218846e-08  0.006057  0.024644  21727.357422   77813.640625  0.104033   
3125  1.102624e-08  0.004972  0.022588  17578.353516   81764.500000  0.106893   
3126  1.761679e-08  0.004887  0.025279  13694.173828   85765.351562  0.118809   

        aod550    aod670    aod865   aod1240      tcco   tc_c2h6  \
0     0.173185  0.133389  0.093220  0.053980  0.000888  0.000003   
1     0.175080  0.136272  0.097289  0.058477  0.000877  0.000003   
2     0.166499  0.128529  0.090249  0.052785  0.000884  0.000003   
3     0.173926  0.133868  0.093414  0.053924  0.000892  0.000003   
4   

In [11]:
final_df = final_df.drop(['location_name', 'NAME'], axis=1)
final_df

fips  year  MeanLifeExpectency       u10       v10         d2m  \
0      1001  2010           75.728489  0.244891 -0.022881  283.480743   
1      1003  2010           77.826608  0.065638  0.113299  286.096985   
2      1005  2010           75.841973  0.336253 -0.075642  284.153351   
3      1007  2010           73.705432  0.221962 -0.004751  283.564850   
4      1009  2010           75.753407  0.284933  0.005598  282.847870   
...     ...   ...                 ...       ...       ...         ...   
3122  56037  2010           77.582422  1.430520  0.814952  269.132935   
3123  56039  2010           83.771012  1.021023  0.729751  268.077728   
3124  56041  2010           77.836526  1.101981  0.647358  270.266602   
3125  56043  2010           78.674180  0.710449  0.532009  270.020416   
3126  56045  2010           79.037741  1.077338  0.246453  271.797424   

             t2m  bcaod550  duaod550       lsm            msl  omaod550  \
0     291.070984  0.006298  0.010663  0.986290  101694.718750  0.076191   
1     292.149780  0.006543  0.013571  0.793853  101686.296875  0.074374   
2     291.171997  0.006456  0.010674  0.987881  101689.679688  0.074089   
3     290.792206  0.006248  0.010553  0.986778  101695.648438  0.077052   
4     289.685669  0.006085  0.008301  0.977253  101711.023438  0.077401   
...          ...       ...       ...       ...            ...       ...   
3122  278.168732  0.004792  0.003537  0.992391  101611.039062  0.043755   
3123  274.648254  0.004971  0.001720  0.982567  101594.773438  0.046533   
3124  277.642670  0.005171  0.002573  0.985319  101640.101562  0.046785   
3125  278.829376  0.005376  0.003680  0.997272  101436.078125  0.050795   
3126  281.669983  0.005429  0.007828  0.998318  101415.359375  0.054489   

               pm1         pm2p5          pm10  ssaod550  suaod550  \
0     1.131367e-08  1.479475e-08  2.029277e-08  0.007522  0.072511   
1     1.082756e-08  1.474964e-08  2.021521e-08  0.010043  0.070549   
2     1.173599e-08  1.520591e-08  2.082824e-08  0.007945  0.067335   
3     1.129947e-08  1.473398e-08  2.021792e-08  0.007260  0.072814   
4     1.163996e-08  1.477842e-08  2.039585e-08  0.006082  0.073962   
...            ...           ...           ...       ...       ...   
3122  5.239014e-09  6.881977e-09  1.016216e-08  0.005140  0.023394   
3123  5.223066e-09  6.534898e-09  9.327799e-09  0.005876  0.021670   
3124  6.811558e-09  8.490385e-09  1.218846e-08  0.006057  0.024644   
3125  5.894271e-09  7.628212e-09  1.102624e-08  0.004972  0.022588   
3126  7.089420e-09  1.112587e-08  1.761679e-08  0.004887  0.025279   

                 z             sp    aod469    aod550    aod670    aod865  \
0      1337.441284  100070.937500  0.210819  0.173185  0.133389  0.093220   
1       426.965088  101166.343750  0.211824  0.175080  0.136272  0.097289   
2      1178.228027  100259.171875  0.202453  0.166499  0.128529  0.090249   
3      1267.733643  100154.929688  0.211800  0.173926  0.133868  0.093414   
4      2292.157227   98932.171875  0.209957  0.171830  0.131437  0.090464   
...            ...            ...       ...       ...       ...       ...   
3122  21153.457031   78338.398438  0.098355  0.080618  0.062090  0.043318   
3123  23587.818359   75929.421875  0.098563  0.080770  0.062109  0.043067   
3124  21727.357422   77813.640625  0.104033  0.085229  0.065549  0.045566   
3125  17578.353516   81764.500000  0.106893  0.087412  0.067119  0.046584   
3126  13694.173828   85765.351562  0.118809  0.097913  0.076147  0.054096   

       aod1240      tcco   tc_c2h6        tchcho   tc_h2o2         tc_oh  \
0     0.053980  0.000888  0.000003  4.347277e-06  0.000009  4.880845e-09   
1     0.058477  0.000877  0.000003  3.707628e-06  0.000009  4.924390e-09   
2     0.052785  0.000884  0.000003  4.157243e-06  0.000009  4.883077e-09   
3     0.053924  0.000892  0.000003  4.335408e-06  0.000009  4.883221e-09   
4     0.050976  0.000888  0.000003  3.997445e-06  0.000008  4.900278e-09   
...    

In [12]:
final_df = final_df.rename(columns=
        {'u10': '10m u-component of wind', 'v10': '10m v-componend of wind', 'd2m':'2m dew point temperature',  
          't2m':'2m temperature', 'bcaod550':'Black carbon AOD at 550 nm','duaod550':'Dust AOD at 550 nm',                                                  
          'lsm':'Land-sea mask', 'msl':'Mean sea level pressure', 'omaod550':'Organic matter AOD at 550 nm',
          'pm1':'PM$_1$', 'pm2p5':'PM$_{2.5}$', 'pm10':'PM$_{10}$', 'ssaod550':'Sea salt AOD at 550 nm',
          'suaod550':'Sulphate AOD at 550 nm', 'z':'Surface geopotential', 'sp':'Surface pressure',  
          'aod469':'Total AOD at 469 nm', 'aod550':'Total AOD at 550 nm', 'aod670':'Total AOD at 670 nm',
          'aod865':'Total AOD at 865 nm', 'aod1240': 'Total AOD at 1240 nm', 'tcco':'Total column carbon monoxide',
          'tc_c2h6':'Total column ethane', 'tchcho':'Total column formaldehyde', 'tc_h2o2':'Total column hydrogen peroxide',
          'tc_oh':'Total column hydroxyl radical', 'tc_c5h8':'Total column isoprene', 'tc_ch4':'Total column methane',
          'tc_hno3':'Total column nitric acid', 'tcno2':'Total column nitrogen dioxide', 'tc_no':'Total column nitrogen monoxide',
          'gtco3':'Total column ozone', 'tc_pan':'Total column peroxyacetyl nitrate','tc_c3h8':'Total column propane',
          'tcso2':'Total column sulphur dioxide', 'tcwv':'Total column water vapour', 'co':'Carbon monoxide',
          'aermr04':'Dust aerosol (0.03-0.55 \u00B5m) mixing ratio', 'aermr05':'Dust aerosol (0.55-0.9 \u00B5m) mixing ratio',
          'aermr06':'Dust aerosol (0.9-20 \u00B5m) mixing ratio', 'c2h6':'Ethane', 'hcho':'Formaldehyde',
          'h2o2':'Hydrogen peroxide', 'aermr09':'Hydrophilic black carbon aerosol mixing ratio', 
          'aermr07':'Hydrophilic organic matter aerosol mixing ratio', 
          'aermr10':'Hydrophobic black carbon aerosol mixing ratio',
          'aermr08':'Hydrophobic organic matter aerosol mixing ratio', 'oh':'Hydroxyl radical',
          'c5h8':'Isoprene', 'hno3':'Nitric acid', 'no2':'Nitrogen dioxide', 'no':'Nitrogen monoxide', 'go3':'Ozone',
          'pan':'Peroxyacetyl nitrate', 'c3h8':'Propane', 'aermr01':'Sea salt aerosol (0.03-0.5 \u00B5m) mixing ratio',
          'aermr02':'Sea salt aerosol (0.5-5 \u00B5m) mixing ratio', 'aermr03':'Sea salt aerosol (5-20 \u00B5m) mixing ratio',
          'q':'Specific humidity', 'aermr11':'Sulphate aerosol mixing ratio', 'so2':'Sulphur dioxide',
          't':'Temperature', 'lai_hv':'Leaf area index, high vegetation', 'lai_lv':'Leaf area index, low vegetation',
          'asn':'Snow albedo', 'sd':'Snow depth', 'tcw':'Total column water',
          'aermssdus':'Vertically integrated mass of dust aerosol (0.03-0.55 \u00B5m)',
         'aermssdum':'Vertically integrated mass of dust aerosol (0.55-9 \u00B5m)',
         'aermssdul':'Vertically integrated mass of dust aerosol (9-20 \u00B5m)',
         'aermssbchphil':'Vertically integrated mass of hydrophilic black carbon aerosol',
         'aermssomhphil':'Vertically integrated mass of hydrophilic organic matter aerosol',
         'aermssbchphob':'Vertically integrated mass of hydrophobic black carbon aerosol',
         'aermssomhphob':'Vertically integrated mass of hydrophobic organic matter aerosol',
         'aermsssss':'Vertically integrated mass of sea salt aerosol (0.03-0.5 \u00B5m)',
         'aermssssm':'Vertically integrated mass of sea salt aerosol (0.5-5 \u00B5m)',
         'aermssssl':'Vertically integrated mass of sea salt aerosol (5-20 \u00B5m)',
         'aermsssu':'Vertically integrated mass of sulphate aerosol',
         'aermssso2':'Vertically integrated mass of sulphur dioxide',         
        })
final_df

fips  year  MeanLifeExpectency  10m u-component of wind  \
0      1001  2010           75.728489                 0.244891   
1      1003  2010           77.826608                 0.065638   
2      1005  2010           75.841973                 0.336253   
3      1007  2010           73.705432                 0.221962   
4      1009  2010           75.753407                 0.284933   
...     ...   ...                 ...                      ...   
3122  56037  2010           77.582422                 1.430520   
3123  56039  2010           83.771012                 1.021023   
3124  56041  2010           77.836526                 1.101981   
3125  56043  2010           78.674180                 0.710449   
3126  56045  2010           79.037741                 1.077338   

      10m v-componend of wind  2m dew point temperature  2m temperature  \
0                   -0.022881                283.480743      291.070984   
1                    0.113299                286.096985      292.149780   
2                   -0.075642                284.153351      291.171997   
3                   -0.004751                283.564850      290.792206   
4                    0.005598                282.847870      289.685669   
...                       ...                       ...             ...   
3122                 0.814952                269.132935      278.168732   
3123                 0.729751                268.077728      274.648254   
3124                 0.647358                270.266602      277.642670   
3125                 0.532009                270.020416      278.829376   
3126                 0.246453                271.797424      281.669983   

      Black carbon AOD at 550 nm  Dust AOD at 550 nm  Land-sea mask  \
0                       0.006298            0.010663       0.986290   
1                       0.006543            0.013571       0.793853   
2                       0.006456            0.010674       0.987881   
3                       0.006248            0.010553       0.986778   
4                       0.006085            0.008301       0.977253   
...                          ...                 ...            ...   
3122                    0.004792            0.003537       0.992391   
3123                    0.004971            0.001720       0.982567   
3124                    0.005171            0.002573       0.985319   
3125                    0.005376            0.003680       0.997272   
3126                    0.005429            0.007828       0.998318   

      Mean sea level pressure  Organic matter AOD at 550 nm        PM$_1$  \
0               101694.718750                      0.076191  1.131367e-08   
1               101686.296875                      0.074374  1.082756e-08   
2               101689.679688                      0.074089  1.173599e-08   
3               101695.648438                      0.077052  1.129947e-08   
4               101711.023438                      0.077401  1.163996e-08   
...                       ...                           ...           ...   
3122            101611.039062                      0.043755  5.239014e-09   
3123            101594.773438                      0.046533  5.223066e-09   
3124            101640.101562                      0.046785  6.811558e-09   
3125            101436.078125                      0.050795  5.894271e-09   
3126            101415.359375                      0.054489  7.089420e-09   

        PM$_{2.5}$     PM$_{10}$  Sea salt AOD at 550 nm  \
0     1.479475e-08  2.029277e-08                0.007522   
1     1.474964e-08  2.021521e-08                0.010043   
2     1.520591e-08  2.082824e-08                0.007945   
3     1.473398e-08  2.021792e-08                0.007260   
4     1.477842e-08  2.039585e-08                0.006082   
...            ...           ...                     ...   
3122  6.881977e-09  1.016216e-08                0.005140   
3123  6.534898e-09  9.327799e-09                0.00587